schism/compiler.ss

```diff
@@ -1096,17 +1107,29 @@
         (cons (cons 'anyref '(ref.null))
               (make-globals-for-literals (- nglobals 1)))))
   ;; -> (fns globals start)
-  (define (lower-literals-in-functions fns)
+  (define (lower-literals-in-definitions fns)
     (let* ((cache '())
            (nglobals 0)
            (start #f)
            (env (cons cache (cons nglobals (cons start '()))))
-           (fns (map (lambda (fn) (lower-literals-in-function fn env)) fns))
+           (fns (map (lambda (fn) (lower-literals-in-definition fn env)) fns))
            (cache (car env))
            (nglobals (cadr env))
-           (start (caddr env))
-           (globals (make-globals-for-literals nglobals)))
-      (cons fns (cons globals (cons start '())))))
+           (fun? (lambda (x) (pair? (car x))))
+           (defs (filter-out fun? fns))
+           (fns (filter fun? fns))
+           (start (fold-left (lambda (start def)
+                               `(seq ,start
+                                     (icall set-global
+                                            (i32 ,(cadr def))
+                                            ,(caddr def))))
+                             (caddr env)
+                             defs))
+           (globals (make-globals-for-literals nglobals))
+           (defs (map (lambda (def)
+                        `(,(car def) . ,(cadr def)))
+                      defs)))
+      `(,fns ,defs ,globals ,start)))
```

Comment:
Also return the top level variables with their reference to allow exporting them